File: server/routes/reviews.js
const express = require('express');
const Review = require('../models/Review');
const User = require('../models/User');
const Project = require('../models/Project');
const auth = require('../middleware/auth');

const router = express.Router();

/**
 * @route   POST /api/reviews
 * @desc    Submit a review for a user after project completion
 * @access  Private
 */
router.post('/', auth, async (req, res) => {
  try {
    const { revieweeId, projectId, rating, comment } = req.body;

    // Check if project exists and is completed (ideal, but checking existence for now)
    const project = await Project.findById(projectId);
    if (!project) {
      return res.status(404).json({ message: 'Project not found' });
    }

    // Basic permission: Reviewer must be part of the project
    const isCompany = project.company.toString() === req.user._id.toString();
    const isFreelancer = project.selectedFreelancer?.toString() === req.user._id.toString();

    if (!isCompany && !isFreelancer) {
      return res.status(403).json({ message: 'You are not authorized to review this project' });
    }

    // Prevent reviewing yourself
    if (revieweeId === req.user._id.toString()) {
      return res.status(400).json({ message: 'You cannot review yourself' });
    }

    // Check if review already exists
    const existingReview = await Review.findOne({
      reviewer: req.user._id,
      reviewee: revieweeId,
      projectId: projectId
    });

    if (existingReview) {
      return res.status(400).json({ message: 'You have already reviewed this user for this project' });
    }

    const review = new Review({
      reviewer: req.user._id,
      reviewee: revieweeId,
      projectId,
      rating,
      comment
    });

    await review.save();
    
    // Populate reviewer info for the response
    await review.populate('reviewer', 'profile.firstName profile.lastName profile.profilePicture');

    res.status(201).json({
      success: true,
      message: 'Review submitted successfully',
      review
    });
  } catch (error) {
    console.error('Review submission error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

/**
 * @route   GET /api/reviews/:userId
 * @desc    Get all reviews for a specific user
 * @access  Private
 */
router.get('/:userId', auth, async (req, res) => {
  try {
    const reviews = await Review.find({ reviewee: req.params.userId })
      .populate('reviewer', 'profile.firstName profile.lastName profile.profilePicture')
      .populate('projectId', 'title')
      .sort({ createdAt: -1 });

    res.json(reviews);
  } catch (error) {
    console.error('Get reviews error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

module.exports = router;

File: server-python/routes/reviews.py
from flask import Blueprint, request, jsonify
from models import User, Project, Review
from extensions import db
from routes.middleware import token_required
from datetime import datetime

reviews_bp = Blueprint('reviews', __name__)

@token_required
def submit_review(current_user):
    data = request.json or {}
    reviewee_id = data.get('revieweeId')
    project_id = data.get('projectId')
    rating = data.get('rating')
    comment = data.get('comment', '')

    if not reviewee_id or not rating:
        return jsonify({"message": "revieweeId and rating are required"}), 400

    if str(reviewee_id) == str(current_user.id):
        return jsonify({"message": "You cannot review yourself"}), 400

    # Check duplicate
    existing = Review.query.filter_by(
        reviewer_id=current_user.id,
        reviewee_id=int(reviewee_id),
        project_id=project_id
    ).first()
    if existing:
        return jsonify({"message": "You have already reviewed this user for this project"}), 400

    review = Review(
        reviewer_id=current_user.id,
        reviewee_id=int(reviewee_id),
        project_id=project_id,
        rating=float(rating),
        comment=comment
    )
    db.session.add(review)

    # Update reviewee rating average
    reviewee = User.query.get(int(reviewee_id))
    if reviewee:
        new_count = reviewee.rating_count + 1
        new_avg = ((reviewee.rating_average * reviewee.rating_count) + float(rating)) / new_count
        reviewee.rating_count = new_count
        reviewee.rating_average = round(new_avg, 2)

    db.session.commit()
    reviewer = current_user.to_dict()
    return jsonify({
        "success": True,
        "message": "Review submitted successfully",
        "review": {
            "_id": str(review.id),
            "rating": review.rating,
            "comment": review.comment,
            "reviewer": {"_id": str(reviewer['_id']), "profile": reviewer['profile']},
            "createdAt": review.created_at.isoformat()
        }
    }), 201

@reviews_bp.route('/<int:user_id>', methods=['GET'])
@token_required
def get_reviews(current_user, user_id):
    reviews = Review.query.filter_by(reviewee_id=user_id)\
                          .order_by(Review.created_at.desc()).all()
    result = []
    for r in reviews:
        reviewer = User.query.get(r.reviewer_id)
        result.append({
            "_id": str(r.id),
            "rating": r.rating,
            "comment": r.comment,
            "createdAt": r.created_at.isoformat(),
            "reviewer": reviewer.to_dict() if reviewer else None
        })
    return jsonify(result), 200
